go/cmd/vtbackup/vtbackup.go

```diff
@@ -14,120 +14,331 @@ See the License for the specific language governing permissions and
+func takeBackup(ctx context.Context, topoServer *topo.Server, backupStorage backupstorage.BackupStorage, backupDir string) error {
+	// This is an imaginary tablet alias. The value doesn't matter for anything,
+	// except that we generate a random UID to ensure the target backup
+	// directory is unique if multiple vtbackup instances are launched for the
+	// same shard, at exactly the same second, pointed at the same backup
+	// storage location.
+	bigN, err := rand.Int(rand.Reader, big.NewInt(math.MaxUint32))
+	if err != nil {
+		return fmt.Errorf("can't generate random tablet UID: %v", err)
+	}
+	tabletAlias := &topodatapb.TabletAlias{
+		Cell: "vtbackup",
+		Uid:  uint32(bigN.Uint64()),
 	}
 
-	dbcfgs, err := dbconfigs.Init(socketFile)
+	// Clean up our temporary data dir if we exit for any reason, to make sure
+	// every invocation of vtbackup starts with a clean slate, and it does not
+	// accumulate garbage (and run out of disk space) if it's restarted.
+	tabletDir := mysqlctl.TabletDir(tabletAlias.Uid)
+	defer func() {
+		log.Infof("Removing temporary tablet directory: %v", tabletDir)
+		if err := os.RemoveAll(tabletDir); err != nil {
+			log.Warningf("Failed to remove temporary tablet directory: %v", err)
+		}
+	}()
+
+	// Start up mysqld as if we are mysqlctld provisioning a fresh tablet.
+	mysqld, mycnf, err := mysqlctl.CreateMysqldAndMycnf(tabletAlias.Uid, *mysqlSocket, int32(*mysqlPort))
 	if err != nil {
-		log.Warning(err)
+		return fmt.Errorf("failed to initialize mysql config: %v", err)
+	}
+	initCtx, initCancel := context.WithTimeout(ctx, *mysqlTimeout)
+	defer initCancel()
+	if err := mysqld.Init(initCtx, mycnf, *initDBSQLFile); err != nil {
+		return fmt.Errorf("failed to initialize mysql data dir and start mysqld: %v", err)
 	}
+	// Shut down mysqld when we're done.
+	defer func() {
+		// Be careful not to use the original context, because we don't want to
+		// skip shutdown just because we timed out waiting for other things.
+		ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+		defer cancel()
+		mysqld.Shutdown(ctx, mycnf, false)
+	}()
 
-	topoServer := topo.Open()
-	mysqld := mysqlctl.NewMysqld(dbcfgs)
-	dir := fmt.Sprintf("%v/%v", *initKeyspace, *initShard)
+	extraEnv := map[string]string{
+		"TABLET_ALIAS": topoproto.TabletAliasString(tabletAlias),
+	}
 
-	pos, err := mysqlctl.Restore(context.Background(), mycnf, mysqld, dir, *concurrency, extraEnv, map[string]string{}, logutil.NewConsoleLogger(), true, dbName)
-	switch err {
-	case nil:
-		// Horray Evenything worked
-		// We have restored a backup ( If one existed, now make sure replicatoin is started
-		if err := resetReplication(context.Background(), pos, mysqld); err != nil {
-			log.Fatalf("Error Starting Replication %v", err)
+	// In initial_backup mode, just take a backup of this empty database.
+	if *initialBackup {
+		// Take a backup of this empty DB without restoring anything.
```

Comment:
Why do we want to take a backup of the empty database?

Reply:
Empty Database isn't a good way to categorize this. The database starts empty, but it will catch up on replication, so it won't be empty when we take the backup. Maybe this should read take a backup starting from an empty database or some thing like that

Reply:
The `initialBackup` mode I'm adding here really does intend to take a backup of an empty database. I'm going to upload this empty backup to the shard's backup location before launching any tablets at all. In this way, I can bootstrap a new shard without ever having to take down any tablet to tell it to take a backup.

Without this, you have to take down a tablet once, when you first deploy the shard. That may not be possible if the user asked for a small number of replicas (e.g. you can't take a backup on a master if they only asked for 1 tablet). Doing it this way also ensures that once the shard is up, it's up for good; we don't have to flap from available to unavailable (to take the initial backup) and then back again, which users might notice if they're starting to test things out. And we can assume going forward that no tablet will ever go down for backup, ever, which simplifies our automation and operations.